File: 146_LRU_cache.java
class LRUCache {
    private class Node {
        int key;
        int value;
        Node prev;
        Node next;
        
        Node(int key, int value){
            this.key = key;
            this.value = value;
            prev = null;
            next = null;
        }
    }
    private int capacity;
    private Map<Integer, Node> map = new HashMap<>();
    private Node head = new Node(-1, -1);
    private Node tail = new Node(-1, -1);
    
    public LRUCache(int capacity) {
        this.capacity = capacity;
        this.head.next = this.tail;
        this.tail.prev = this.head;
    }
    
    public int get(int key) {
        if (! map.containsKey(new Integer(key))) return -1;
        Node curr = map.get(key);
        curr.prev.next = curr.next;
        curr.next.prev = curr.prev;

        moveToTail(curr);

        return map.get(key).value;
    }
    
    public void put(int key, int value) {
        if (get(key) != -1) {
            map.get(key).value = value;
            return;
        }
        if (map.size() == capacity) {
            map.remove(head.next.key);
            head.next = head.next.next;
            head.next.prev = head;
        }
        Node insert = new Node(key, value);
        map.put(key, insert);
        moveToTail(insert);
    }
    
    private void moveToTail(Node curr) {
        curr.next = tail;
        tail.prev.next = curr;
        curr.prev = tail.prev;
        tail.prev = curr;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: python/146_LRU_cache.py
class Node:
    def __init__(self, key: str, value: str):
        self.key = key
        self.value = value
        self.prev = None
        self.next = None


class LRUCache:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.map = dict()

        self.head = Node(-1, -1)
        self.tail = Node(-1, -1)

        self.head.next = self.tail
        self.tail.prev = self.head

    def get(self, key: int) -> int:
        # if not in map, return -1
        if key not in self.map: return -1

        # key in map, move the node to the end of list
        curr = self.map[key]
        curr.prev.next = curr.next
        curr.next.prev = curr.prev

        self.moveToTail(curr)

        return curr.value

    def put(self, key: int, value: int) -> None:
        if self.get(key) != -1:
            self.map[key].value = value
            return

        if self.capacity == len(self.map):
            # evict the least used one in the head
            del self.map[self.head.next.key]
            self.head.next = self.head.next.next
            self.head.next.prev = self.head

        insert = Node(key, value)
        self.map[key] = insert
        self.moveToTail(insert)

    def moveToTail(self, curr: Node):
        curr.next = self.tail
        self.tail.prev.next = curr

        curr.prev = self.tail.prev
        self.tail.prev = curr

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
